mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp

```diff
@@ -125,9 +126,15 @@ class BufferizeBackpropOp : public OpConversionPattern<BackpropOp> {
         // Enzyme requires buffers for the primal outputs as well, even though we don't need their
         // values. We'll mark them dupNoNeed later on to allow Enzyme to optimize away their
         // computation.
-        SmallVector<Value> calleeResults;
-        ValueRange resShadows = adaptor.getCotangents();
-        generateAllocations(rewriter, loc, calleeResults, resShadows);
+        SmallVector<Value> calleeResults, resShadows;
+        ValueRange cotangents = adaptor.getCotangents();
+        generateAllocations(rewriter, loc, calleeResults, cotangents);
+        // Enzyme mutates the result shadows but the cotangent tensors must be immutable, so we
+        // create copies to pass into Enzyme.
+        generateAllocations(rewriter, loc, resShadows, cotangents);
+        for (const auto &[cotangent, resShadow] : llvm::zip(cotangents, resShadows)) {
+            rewriter.create<memref::CopyOp>(loc, cotangent, resShadow);
+        }
```

Comment:
So when you have a primal in destination passing style, you store to its output. For instance:
```
func @primal(x: f64, out: ptr<f64>) {
  store x, out : f64 into ptr<f64>
}
```

When differentiating this with an AD system that supports mutation, the correct way to handle a store involves zeroing out the derivative of the store's destination. This is because you've written to the store's destination, so any gradient value that was previously there should be erased by zeroing it out. [You can see this rule in the Enzyme GPU paper (page 2, top-right figure where it shows the "Memory store" and "Reverse memory store")](https://dl.acm.org/doi/pdf/10.1145/3458817.3476165):

```
func @adjoint(x: f64, out: ptr<f64>, dout: ptr<f64>) -> f64 {
  dx = alloca : ptr<f64>;
  store 0.0, dx : f64 into ptr<f64>

  tmp = load dout : ptr<f64>
  store 0.0, dout : f64 into ptr<f64>
  load/store dx += tmp
  return load dx
}
```

However, if we have a program that calls grad multiple times, I found that `one-shot-bufferize=dialect-filter=memref` will actually run CSE assuming the cotangent tensors are immutable. The gradient lowering will generate this:
```mlir
%empty = tensor.empty() : tensor<f64>
%cotangent0 = tensor.insert 1.0 into %empty[] : tensor<f64>
gradient.backprop ... cotangent(%cotangent0) ...
%cotangent1 = tensor.insert 1.0 into %empty[] : tensor<f64>
gradient.backprop ... cotangent(%cotangent1) ...
```

The one shot bufferize will transform it into:
```mlir
%empty = tensor.empty() : tensor<f64>
%cotangent = tensor.insert 1.0 into %empty[] : tensor<f64>
gradient.backprop ... cotangent(%cotangent) ...
gradient.backprop ... cotangent(%cotangent) ...
```

The first backprop call will produce the correct result, but the second one will incorrectly produce a zero gradient because Enzyme stores a zero to the (shared) cotangent tensor unless we explicitly bufferize to enforce that the inputs to the backprop op are immutable.

Reply:
Oh interesting, yeah that's super important thanks for the explanation! Updating the comment with the reason for the copy (one-shot-bufferize CSEing the cotangent vectors) would be useful.

I'm guessing this is a quirk of our bufferization pipeline not being handled in a unified way or being abused beyond what MLIR supports 🤔

Reply:
I would more say that every per-dialect bufferization pass is implemented to conservatively assume any DPS output could be mutated once bufferized, therefore new allocations must be made for all of them. Finding cases where the output is no longer used and thus safe to be used without a new alloc + copy is exactly the job of one shot bufferization.